pkg/network/vmispec: Add linter coverage

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
- Break long lines
- Apply gofumpt
- Remove redundant newlines
- Fix typos
- Convert snake_case to camelCase

Added a case to `SetDefaultNetworkInterface` in order to also consider `v1.DeprecatedPasstInterface`.

Propose an alternative implementation for `PopInterfaceByNetwork`.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer
<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

orelmisan commented on pkg/network/vmispec/interface.go:
```diff
@@ -102,7 +109,10 @@ func IsPodNetworkWithBridgeBindingInterface(networks []v1.Network, ifaces []v1.I
+func PopInterfaceByNetwork(
+	statusIfaces []v1.VirtualMachineInstanceNetworkInterface,
+	network *v1.Network,
+) (*v1.VirtualMachineInstanceNetworkInterface, []v1.VirtualMachineInstanceNetworkInterface) {
```
The linter fails on this line with the following error message:
> pkg/network/vmispec/interface.go:112:1: unnamedResult: consider giving a name to these results (gocritic)
func PopInterfaceByNetwork(

orelmisan commented:
/uncc @xpivarc @RamLavi 
/cc @nirdothan @ormergi

orelmisan commented:
> Thanks @orelmisan Please see my comments. one more comment that is indirectly related:
> 
> https://github.com/kubevirt/kubevirt/blob/1593dae0378b81f1dd82976f61fd10a944741864/pkg/network/vmispec/defaults.go#L28
> 
> netClusterConfiger->netClusterConfigurer

Thank you for the review @nirdothan.
I think this one is out of scope for this PR.
There will be a PR to graduate the `DynamicPodInterfaceNaming` FG and the said interface could be removed there.

orelmisan commented on pkg/network/setup/netstat.go:
```diff
@@ -161,6 +158,27 @@ func restorePodIfaceNameForPrimaryIface(
+		return interfacesStatus
+	}
+
+	ifacesStatusCopy := slices.Clone(interfacesStatus)
```
> why do you need to copy the slice?
1. To make sure the function does not modify the input.
2. To emphasize the function creates a new value based on the input.

> why not just swap locations interfacesStatus[primaryIfaceStatusIndex] <-> interfacesStatus[0] ?

Good idea! Done.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/edad862558e2e45d248fcdf591c79aeb3fac4cf0..fe4b96754aca3047b8a1b1005a08d85f4f9e05f5): Answered @nirdothan's [comment](https://github.com/kubevirt/kubevirt/pull/13153#pullrequestreview-2430418729).

EdDev commented on pkg/network/vmispec/interface_test.go:
```diff
@@ -91,41 +91,6 @@ var _ = Describe("VMI network spec", func() {
 
-	const iface1, iface2, iface3, iface4, iface5 = "iface1", "iface2", "iface3", "iface4", "iface5"
-
-	Context("pop interface by network", func() {
```
Are these scenarios covered at an upper level?

orelmisan replied:
I've dropped this change in order to limit the scope of this PR.

EdDev replied:
ping

orelmisan replied:
This scenario should be covered by:
https://github.com/kubevirt/kubevirt/blob/a3b5ee1fc5eddd4a31ec29cf6493a7834cbda5ff/pkg/network/setup/netstat_test.go#L625

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/13ad4e26b04fd3a99584b73f81a34aaa10c4e4a2..bef6257e01c307192ce68112513f20b188e6239b): 
1. Rebase
2. Split the linter coverage commit.
3. Dropped the change to NetStat.

orelmisan commented on pkg/network/setup/netstat.go:
```diff
@@ -161,6 +158,25 @@ func restorePodIfaceNameForPrimaryIface(
+	}
+
+	ifacesStatusCopy := slices.Clone(interfacesStatus)
+	ifacesStatusCopy[0], ifacesStatusCopy[primaryIfaceStatusIndex] = ifacesStatusCopy[primaryIfaceStatusIndex], ifacesStatusCopy[0]
```
I've dropped this change in order to limit the scope of this PR.

EdDev replied:
ping

orelmisan replied:
Adjusted the logic to match the previous implementation - the primary network is moved to the front of the list, and the rest of the elements keep their ordering.

orelmisan commented:
/kind cleanup

orelmisan commented:
> thanks @orelmisan. Looks good. Why do you call it "Linter coverage" ?

@nirdothan Thank you for reviewing the PR.
Following this PR, the linter covers / protects the `pkg/network/vmispec` package.
Any change that causes the linter to fail - will block the PR proposing it.

EdDev commented on pkg/network/vmispec/defaults.go:
```diff
@@ -50,6 +50,8 @@ func SetDefaultNetworkInterface(config netClusterConfiger, spec *v1.VirtualMachi
 		spec.Domain.Devices.Interfaces = []v1.Interface{*v1.DefaultMasqueradeNetworkInterface()}
 	case v1.DeprecatedSlirpInterface:
 		return fmt.Errorf("slirp interface is deprecated as of v1.3")
+	case v1.DeprecatedPasstInterface:
+		return fmt.Errorf("passt interface is deprecated as of v1.3")
```
Slirp is handled here because it was not protected using a FG.
Passt on the other hand was enabled through a FG, therefore the warning is in the FG context [1].

[1] https://github.com/kubevirt/kubevirt/blob/a3b5ee1fc5eddd4a31ec29cf6493a7834cbda5ff/pkg/virt-config/deprecation/feature-gates.go#L106

orelmisan replied:
The linter failed here with the following error message:
```
 missing cases in switch of type v1.NetworkInterfaceType: v1.DeprecatedPasstInterface (exhaustive)
        switch v1.NetworkInterfaceType(config.GetDefaultNetworkInterface()) 
```

The alternative is to silence the linter.
WDYT?

EdDev replied:
I think the linter asks to be explicit about not treating other options.
Does adding `default: // do nothing` satisfy the linter?

orelmisan replied:
> I think the linter asks to be explicit about not treating other options. Does adding `default: // do nothing` satisfy the linter?

No:
```
pkg/network/vmispec/defaults.go:43:2: missing cases in switch of type v1.NetworkInterfaceType: v1.DeprecatedPasstInterface (exhaustive)
        switch v1.NetworkInterfaceType(config.GetDefaultNetworkInterface()) {
```

orelmisan replied:
BTW, this code snippet is responsible for adding the default pod network with the cluster-wide default binding and is called before the FG check.

EdDev replied:
Then a noop on that case?
I fail to understand why it does not complain about macvtap then.

And it does not make a lot of sense in general. If we had 20 options, do we need to express them all? Something is wrong here.

orelmisan replied:
It examines all the options of `NetworkInterfaceType`:

EdDev replied:
Choose one of the options:
- Add it with a noop.
- Update the linter config to supress it in case a default is defined: https://golangci-lint.run/usage/linters/#exhaustive
- Remove it from that const group and mimic the old macvtap.

orelmisan replied:
I am in favor of removing the option from the const group, I think it also makes sense to remove slirp as well (not necessarily in this PR).
WDYT?

EdDev replied:
Fine by me.
But slirp is a special case, it got discontinued from ga with an exception. I would not touch it to express that.

orelmisan replied:
Done.

EdDev commented on pkg/network/vmispec/interface.go:
```diff
@@ -102,7 +109,11 @@ func IsPodNetworkWithBridgeBindingInterface(networks []v1.Network, ifaces []v1.I
 }
 
-func PopInterfaceByNetwork(statusIfaces []v1.VirtualMachineInstanceNetworkInterface, network *v1.Network) (*v1.VirtualMachineInstanceNetworkInterface, []v1.VirtualMachineInstanceNetworkInterface) {
+//nolint:gocritic
```
Sorry, but I feel that if I let one of these to get in, it will be a precedence to other exceptions.
I would prefer to "eat" the named returned arguments over adding such a linter override.

orelmisan replied:
I've re-added the removal of `PopInterfaceByNetwork`.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/bef6257e01c307192ce68112513f20b188e6239b..a5b11633d890ea4917766677de6bcc9fd474415b): Addressed @EdDev's [comment](https://github.com/kubevirt/kubevirt/pull/13153#discussion_r1896742245). 
Re-added a commit to rename the `netClusterConfiger` interface that was mistakenly removed after the rebase.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/a5b11633d890ea4917766677de6bcc9fd474415b..362bb20b94c3d5677e17feba4b91c307f31b1f2e): Addressed @EdDev's [comment](https://github.com/kubevirt/kubevirt/pull/13153#discussion_r1897140964).

orelmisan commented:
Linter [error](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/13153/pull-kubevirt-code-lint/1871813093047144448) is unrelated.

/test pull-kubevirt-code-lint

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/362bb20b94c3d5677e17feba4b91c307f31b1f2e..b91b78a4b8244cc19ce7d97d8f3954cb7a0cdcba): Addressed @EdDev's [comment](https://github.com/kubevirt/kubevirt/pull/13153#discussion_r1846445057).

EdDev reviewed (commented):
Thank you!

/approve

orelmisan commented:
> I think that the diff link [here](https://github.com/kubevirt/kubevirt/pull/13153#issuecomment-2562282047) is wrong

Why do you think so?
I doubled check and the change addresses the comment.

orelmisan commented on pkg/network/setup/netstat.go:
```diff
@@ -155,6 +153,29 @@ func restorePodIfaceNameForPrimaryIface(
 	return interfacesStatus
 }
 
+func movePrimaryIfaceStatusToFront(
```
> I'm not sure what went on here. You've replaced PopInterfaceByNetwork with movePrimaryIfaceStatusToFront. How does it relate to linting?

Please see the commit message of [net, vmispec: Add linter coverage](https://github.com/kubevirt/kubevirt/pull/13153/commits/bf4c74736f299646c8c569fddb7cde8a1579b4f9).

orelmisan replied:
I've updated the commit message.

enp0s3 replied:
@orelmisan Thanks for the explanation. Can you please mention that you've changed the function signature in order to make linter happy, while the functionality remains the same, and protected by the current unit-tests, so despite it looks like some new code was added, actually its a refactoring.

orelmisan replied:
> Can you please mention that you've changed the function signature in order to make linter happy, while the functionality remains the same, and protected by the current unit-tests, so despite it looks like some new code was added, actually its a refactoring.

I've adjusted the commit message a bit to clarify.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/b91b78a4b8244cc19ce7d97d8f3954cb7a0cdcba..c483f5dbf9ef02527578e1b1e341ae1bbd1d2497): Addressed @nirdothan's [comment](https://github.com/kubevirt/kubevirt/pull/13153#discussion_r1899110844).

enp0s3 commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -1798,9 +1798,6 @@ const (
 	DeprecatedSlirpInterface NetworkInterfaceType = "slirp"
 	// Virtual machine instance masquerade interface
 	MasqueradeInterface NetworkInterfaceType = "masquerade"
-	// Virtual machine instance passt interface is deprecated
```
@orelmisan I see that SLIRP and PASST have both the same comment:
```
// Deprecated: Removed in v1.3.
```
I don't understand why SLIRP comparing to PASST wasn't removed.

orelmisan replied:
Thank you for reviewing the PR @enp0s3.
SLIRP was discontinued after it was GA.
Passt was discontinued when it still had a feature gate protecting it.
Please see https://github.com/kubevirt/kubevirt/pull/13153#discussion_r1897142806.

enp0s3 reviewed (commented):
@orelmisan Hi! thank you for the effort!

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/c483f5dbf9ef02527578e1b1e341ae1bbd1d2497..5b7139465bebe38baed21569febe0f649332aefb): Rebase
Simplified `movePrimaryIfaceStatusToFront`.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/5b7139465bebe38baed21569febe0f649332aefb..fb4b2727536e49e3de280f1576930f6b1462687a): Rebase.

enp0s3 commented:
@orelmisan Thank you!
/approve
